pilot/pkg/proxy/envoy/v2/internalgen.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	gogoproto "github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+)
+
+const (
+	TypeURLConnections = "istio.io/connections"
+	TypeURLDisconnect  = "istio.io/disconnect"
+
+	// TODO: TypeURLReady - readiness events for endpoints, agent can propagate
+
+	// TypeURLNACK will receive messages of type DiscoveryRequest, containing
+	// the 'NACK' from envoy on rejected configs. Only ID is set in metadata.
+	// This includes all the info that envoy (client) provides.
+	TypeURLNACK = "istio.io/nack"
+)
+
+// InternalGen is a Generator for XDS status updates: connect, disconnect, nacks, acks
+type InternalGen struct {
+	Server *DiscoveryServer
+
+	// TODO: track last N Nacks and connection events, with 'version' based on timestamp.
+	// On new connect, use version to send recent events since last update.
+}
+
+func (sg *InternalGen) OnConnect(node gogoproto.Message) {
+	sg.startPush(TypeURLConnections, []*any.Any{util.MessageToAny(node)})
+}
+
+func (sg *InternalGen) OnDisconnect(node gogoproto.Message) {
+	sg.startPush(TypeURLDisconnect, []*any.Any{util.MessageToAny(node)})
+}
+
+func (sg *InternalGen) OnNack(node *model.Proxy, dr *xdsapi.DiscoveryRequest) {
+	// Make sure we include the ID - the DR may not include metadata
+	dr.Node.Id = node.ID
+	sg.startPush(TypeURLNACK, []*any.Any{util.MessageToAny(dr)})
+}
+
+// startPush is similar with DiscoveryServer.startPush() - but called directly,
+// since status discovery is not driven by config change events.
+// We also want connection events to be dispatched as soon as possible,
+// they may be consumed by other instances of Istiod to update internal state.
+func (sg *InternalGen) startPush(typeURL string, data []*any.Any) {
```

Comment:
Should there be unit tests for these methods?

Reply:
There is a unit test in the PR - it checks that an initial response is sent and that on a new connection there is a push with the event.

Reply:
And more tests will be added as it gets used - for now it's experimental, we need to find out the right interface based on the various use cases. John is using this for the config-less multi cluster, it may be used for debug - and I'm planning to use it for some optimizations on discovery propagation latency.